common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Extensions/DataProtectionBuilderEx.cs

```diff
@@ -0,0 +1,218 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.IIoT.Services.Auth {
+    using Microsoft.AspNetCore.DataProtection;
+    using Microsoft.Azure.IIoT.Auth.Clients;
+    using Microsoft.Azure.IIoT.Crypto.KeyVault;
+    using Microsoft.Azure.IIoT.Crypto.KeyVault.Runtime;
+    using Microsoft.Azure.IIoT.Storage.Blob;
+    using Microsoft.Azure.IIoT.Storage.Blob.Runtime;
+    using Microsoft.Azure.IIoT.Utils;
+    using Microsoft.Azure.KeyVault;
+    using Microsoft.Azure.KeyVault.Models;
+    using Microsoft.Azure.KeyVault.WebKey;
+    using Microsoft.Azure.Services.AppAuthentication;
+    using Microsoft.Azure.Storage;
+    using Microsoft.Azure.Storage.Blob;
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.DependencyInjection;
+    using Serilog;
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Add data protection using azure blob storage and keyvault
+    /// </summary>
+    public static class DataProtectionBuilderEx {
+
+        /// <summary>
+        /// Add azure data protection
+        /// </summary>
+        /// <param name="services"></param>
+        /// <param name="configuration"></param>
+        public static void AddAzureDataProtection(
+            this IServiceCollection services, IConfiguration configuration = null) {
+            if (configuration == null) {
+                configuration = services.BuildServiceProvider()
+                    .GetRequiredService<IConfiguration>();
+            }
+            services.AddDataProtection()
+                .AddAzureBlobKeyStorage(configuration)
+                .AddAzureKeyVaultDataProtection(configuration);
+        }
+
+        /// <summary>
+        /// Add Keyvault protection
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="configuration"></param>
+        public static IDataProtectionBuilder AddAzureKeyVaultDataProtection(
+            this IDataProtectionBuilder builder, IConfiguration configuration = null) {
+            if (configuration == null) {
+                configuration = builder.Services.BuildServiceProvider()
+                    .GetRequiredService<IConfiguration>();
+            }
+            var config = new DataProtectionConfig(configuration);
+            if (string.IsNullOrEmpty(config.KeyVaultBaseUrl)) {
+                return builder;
+            }
+            var keyName = "dataprotection";
+            var client = TryKeyVaultClientAsync(config.KeyVaultBaseUrl,
+                config, keyName).Result;
```

Comment:
Why can await not be used here?

Reply:
Yes, that would be nice, but convention adding services to DI is synchronous call.